opamp/observiq/observiq_client.go

```diff
@@ -0,0 +1,190 @@
+
+	observiqClient := &Client{
+		logger:        clientLogger,
+		ident:         newIdentity(clientLogger, args.Config),
+		configManager: configManager,
+		collector:     args.Collector,
+		endpoint:      args.Config.Endpoint,
+		secretKey:     args.Config.GetSecretKey(),
+	}
+
+	// Parse URL to determin scheme
+	opampURL, err := url.Parse(args.Config.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create collect client based on URL scheme
+	switch opampURL.Scheme {
+	case "ws", "wss":
+		observiqClient.opampClient = client.NewWebSocket(clientLogger)
+	default:
+		return nil, ErrUnsupportedURL
+	}
+
+	// Add configs to config manager
+	configManager.AddConfig(ManagerConfigName, &opamp.ManagedConfig{
+		ConfigPath: args.ManagerConfigPath,
+		Reload:     managerReload(observiqClient, args.ManagerConfigPath),
+	})
+
+	configManager.AddConfig(CollectorConfigName, &opamp.ManagedConfig{
+		ConfigPath: args.CollectorConfigPath,
+		Reload:     collectorReload(observiqClient, args.CollectorConfigPath),
+	})
+
+	configManager.AddConfig(args.LoggerConfigPath, &opamp.ManagedConfig{
+		ConfigPath: args.LoggerConfigPath,
+		Reload:     loggerReload(observiqClient, args.LoggerConfigPath),
+	})
+
+	return observiqClient, nil
+}
+
+// Connect initiates a connection to the OpAmp server
+func (c *Client) Connect(ctx context.Context) error {
+	// Compose and set the agent description
+	if err := c.opampClient.SetAgentDescription(c.ident.ToAgentDescription()); err != nil {
+		c.logger.Error("Error while setting agent description", zap.Error(err))
+		return err
+	}
+
+	settings := types.StartSettings{
+		OpAMPServerURL:      c.endpoint,
+		AuthorizationHeader: c.secretKey,
+		TLSConfig:           nil, // TODO add support for TLS
+		InstanceUid:         c.ident.agentID,
+		Callbacks: types.CallbacksStruct{
+			OnConnectFunc:          c.onConnectHandler,
+			OnConnectFailedFunc:    c.onConnectFailedHandler,
+			OnErrorFunc:            c.onErrorHandler,
+			OnRemoteConfigFunc:     c.onRemoteConfigHandler,
+			GetEffectiveConfigFunc: c.onGetEffectiveConfigHandler,
+			// Unimplemented Handles
+			// SaveRemoteConfigStatusFunc:
+			// OnOpampConnectionSettingsFunc
+			// OnOpampConnectionSettingsAcceptedFunc
+			// OnOwnTelemetryConnectionSettingsFunc
+			// OnOtherConnectionSettingsFunc
+			// OnPackagesAvailableFunc
+			// OnAgentIdentificationFunc
+			// OnCommandFunc
+		},
+	}
+
+	// Start the embedded collector
+	if err := c.collector.Run(ctx); err != nil {
+		return fmt.Errorf("collector failed to start: %w", err)
+	}
+
+	return c.opampClient.Start(ctx, settings)
+}
+
+// Disconnect disconnects from the server
+func (c *Client) Disconnect(ctx context.Context) error {
+	c.collector.Stop()
+	return c.opampClient.Stop(ctx)
+}
+
+// client callbacks
+
+func (c *Client) onConnectHandler() {
+	c.logger.Info("Successfully connected to server")
+
+	if err := c.opampClient.SetAgentDescription(c.ident.ToAgentDescription()); err != nil {
```

Comment:
@cpheps is this here because configs could change now during the opamp lifecycle of the collector?

Reply:
actually that's a good point we don't need to set this in the onConnect as we do it in the Connection function. I'll remove it. You only need to `SetAgentDescription` when it changes. So we only do that when we first start and on any changes to the manager.yml

Reply:
Just pushed an update that removed this